### ISSUE TITLE
Question about divergence/entropy script

### ISSUE THREAD

**Issue by ryapark** [alecwangcq/f-divergence-dpo#1]
Hi, thank you for providing the code for your paper. In your `imbd_eval_metrics.py` script, why do you compute entropy as mean of `P(x) * log P(x)`? Since entropy is expectation of -log P(x) why is it not just mean of `-log P(x)`? I have the same question about the f-divergence computation.

Additionally, in the same script, the `get_f_divergence` function seems to not match the table in the paper detailing the various f-divergences and their generators `f`?
```python
def get_f_divergence(f_divergence, alpha=0.5):
    if f_divergence == 'reverse_kl':
        f = lambda x: -torch.log(x)
        f_prime_one = -1.
    elif f_divergence == 'forward_kl':
        f = lambda x: x * torch.log(x)
        f_prime_one = 1.
```

Thank you and apologies if I am misunderstanding or missing something small.

**Comment by alecwangcq** [comment]
Hi,

>  compute entropy 

Please refer to this line: https://github.com/alecwangcq/f-divergence-dpo/blob/main/metrics/imdb/imdb_eval_metrics.py#L162, I computed the token-wise KL using (\sum_t \int -p(x_t|x_{0:t-1}) \log p(x_t|x_{0:t-1})) / T. 

> get_f_divergence

Thank you for your attention to detail. I reversed the positions of  p and  q in the paper to align with the reward (f'(u)) being consistent with the implicit reward in the DPO paper for reverse KL. The interpretation of forward KL and reverse KL depends on which distribution is considered the variational/approximate one.

However, in the implementation, I followed the conventional definitions, see: https://en.wikipedia.org/wiki/F-divergence#Common_examples_of_f-divergences

Sorry for the confusion.

**Comment by ryapark** [comment]
Thanks for your reply. I am still confused by the entropy computation, since in the code it looks like you are averaging $$-\sum_t p(x_t | x_{<t}) \log p(x_t | x_{<t})$$ over all batches. However, shouldn't the entropy be $$E[-\log p(x_{0:t})]=E[-\sum_t \log p(x_t | x_{<t})]$$ instead of $$E[-\sum_t p(x_t | x_{<t}) \log p(x_t | x_{<t})]$$?

Apologies again if I'm missing something.

**Comment by alecwangcq** [comment]
I believe your understanding is mostly correct, but there may be a misunderstanding of what I computed. This line of code calculates the averaged **per-token entropy**. What you mentioned is calculating the entropy over sequences, which is quite difficult to compute exactly. 

Below is a breakdown of what that line of code does, in case you'd like more context or clarification.

-----

 ```python 
batch_entropy = (-((model_log_probs.exp() * model_log_probs).sum(-1)) * attention_mask).sum() / attention_mask.sum()
```

Firstly, `model_log_probs` is a `batch_size x T x V` tensor, where T is the maximal sequence length, and V is the vocab size.

1. `(model_log_probs.exp() * model_log_probs).sum(-1)`:  this returns a tensor of shape `batch_size x T`, where each element is $\mathbb{H}(p(x_t|x_{< t})) = \sum_{x_t \in V} -p(x_t|x_{<t}) \log p(x_t|x_{<t})$  --> this is the entropy for each time-step, or the token-wise entropy.
2. `(model_log_probs.exp() * model_log_probs).sum(-1) * attention_mask`: this step is for masking out those padding tokens, which will not contribute to the computation of averaged per-token entropy.
3. `(-((model_log_probs.exp() * model_log_probs).sum(-1)) * attention_mask).sum() / attention_mask.sum()`: this step gives the averaged per-token entropy.

**Comment by ryapark** [comment]
I see, I thought you were computing sequence-level entropy. Thank you for the clarification!

One last question - why is computing the entropy over entire sequences difficult? Couldn't you take the average of $$-\sum_t \log p(x_t | x_{<t})$$ to estimate sequence-level entropy? Or are you saying it's easy to approximate but difficulty lies in exact computation?

**Comment by alecwangcq** [comment]
Yes. But, the summation over all the sequences of length T is intractable,
when T is large. It’s easy to approximate but difficult to compute exactly.

**Comment by ryapark** [comment]
Thank you for the clarifications! Closing now
